lib/sqlalchemy/dialects/sqlite/base.py

```diff
@@ -402,6 +402,253 @@ def set_sqlite_pragma(dbapi_connection, connection_record):
+    do_update_stmt_A = insert_stmt.on_conflict_do_update(
+        constraint=unique_constraint_pk,
+        set_=dict(id=1, data='updated value')
+    )
+
+    do_update_stmt_B = insert_stmt.on_conflict_do_update(
+        constraint='id',
+        set_=dict(id=1, data='updated value')
+    )
+
+* The :paramref:`.Insert.on_conflict_do_update.constraint` argument may
+  also refer to a SQLAlchemy construct representing a constraint,
+  e.g. :class:`.UniqueConstraint` or :class:`.PrimaryKeyConstraint`.
+  In this use, only the name of the constraint's first column is used directly.
+  If the constraint is unnamed, then inference will be used,
+  where the expressions and optional WHERE clause of the constraint will
+  be spelled out in the construct::
+
+    unique_constraint = schema.UniqueConstraint(
+        my_table.c.user_email, name="uq_user_email"
+    )
+
+    do_update_stmt = insert_stmt.on_conflict_do_update(
+        constraint=unique_constraint,
+        set_=dict(data='updated value')
+    )
+
+``ON CONFLICT...DO UPDATE`` is used to perform an update of the already
+existing row, using any combination of new values as well as values
+from the proposed insertion.   These values are specified using the
+:paramref:`.Insert.on_conflict_do_update.set_` parameter.  This
+parameter accepts a dictionary which consists of direct values
+for UPDATE::
+
+    from sqlalchemy.dialects.sqlite import insert
+
+    stmt = insert(my_table).values(id='some_id', data='inserted value')
+    do_update_stmt = stmt.on_conflict_do_update(
+        index_elements=['id'],
+        set_=dict(data='updated value')
+        )
+    conn.execute(do_update_stmt)
+
+.. warning::
+
+    The :meth:`_expression.Insert.on_conflict_do_update`
+    method does **not** take into
+    account Python-side default UPDATE values or generation functions, e.g.
+    those specified using :paramref:`_schema.Column.onupdate`.
+    These values will not be exercised for an ON CONFLICT style of UPDATE,
+    unless they are manually specified in the
+    :paramref:`.Insert.on_conflict_do_update.set_` dictionary.
+
+In order to refer to the proposed insertion row, the special alias
+:attr:`~.sqlite.Insert.excluded` is available as an attribute on
+the :class:`_sqlite.Insert` object; this object creates an "excluded." prefix
+on a column, that informs the DO UPDATE to update the row with the value that
+would have been inserted had the constraint not failed::
+
+    from sqlalchemy.dialects.sqlite import insert
+
+    stmt = insert(my_table).values(
+        id='some_id',
+        data='inserted value',
+        author='jlh')
+    do_update_stmt = stmt.on_conflict_do_update(
+        index_elements=['id'],
+        set_=dict(data='updated value', author=stmt.excluded.author)
+        )
+    conn.execute(do_update_stmt)
+
+The :meth:`_expression.Insert.on_conflict_do_update` method also accepts
+a WHERE clause using the :paramref:`.Insert.on_conflict_do_update.where`
```

Comment:
**mike bayer** (zzzeek) wrote:

header:

Limiting Update Rows
^^^^^^^^^^^^^^^^^^^^^


long docs!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2230